Spike: add Puppeteer tests

microbit-rosslowe opened the pull request:
Integrate Puppeteer tests into the existing Python Editor tests.

Includes:
- InvalidTest: Tests that attempting to load a MakeCode hex file will show an error dialog
- OldTestA: Tests that a v1.0.1 hex file loads correctly
- OldTestB: Tests that a v0.9 hex file loads correctly
- URLTest: Tests that loading a project via URL will load the correct content in the editor.

microbit-rosslowe committed:
Spike: add Puppeteer tests

Integrate Puppeteer tests into the existing Python Editor tests.

microbit-rosslowe commented:
In `./tests` run `npm i` then `npm test`.

microbit-rosslowe committed:
Make tests more strict

microbit-rosslowe commented:
Move existing tests -> each test should have its own `-spec` file instead of being integrated into `python-spec.js`.

microbit-rosslowe committed:
Restructure tests

- restructure tests so each test has its own -spec.js file
- start & stop local server after all tests
- remove device aspect of tests, and remove dependencies

microbit-rosslowe commented:
@microbit-carlos I've restructured the tests in https://github.com/microbit-rosslowe/PythonEditor/pull/9/commits/25c8f5940498eb812c8c5c88ad3f141c9ae78769

microbit-rosslowe committed:
Reorganise tests; fix file paths

Neaten & fix tests by fixing test file paths.

microbit-rosslowe committed:
Move Puppeteer tests to a single file

Allows us to use a single browser object across all tests.

microbit-rosslowe commented:
@microbit-carlos I've updated the tests so that the Puppeteer tests all belong in a single `-spec.js` file. This allows for using a single `browser` object across all tests, and also makes it clear how each test works and where the setup & teardown is. I've also cleaned up the tests so they should all (hopefully!) follow AAA better.

microbit-rosslowe committed:
No need to remove line

microbit-carlos commented on tests/spec/puppeteer-spec.js:
```diff
@@ -0,0 +1,98 @@
+        global.browser.close();
+    });
+
+    it("Correctly loads a script from an import URL.", async function() {
```
Can we also check the script name in this test?

microbit-rosslowe replied:
There's a future task to check the script name doesn't change when loading an invalid file (i.e. the Makecode test). In the current version of the editor these tests are against, this test will fail.

microbit-carlos replied:
This is for the import URL, do you mean that when loading a script via URL the script name doesn't change?

microbit-rosslowe replied:
I'm adding script name checks to each test, this was more of a generic comment. No; all other tests change the name correctly.

microbit-carlos replied:
Okay, so are you adding the name checks in this PR? I'm a bit confused as to when that will happen.

microbit-rosslowe replied:
Changed in https://github.com/microbit-rosslowe/PythonEditor/pull/9/commits/3dcdaf967f1b0b1ec45374b6e936c721b0e0781d

microbit-carlos commented on tests/spec/puppeteer-spec.js:
```diff
@@ -0,0 +1,98 @@
+
+        let hasLoadedContent = false;
+        const codeContent = await page.evaluate("window.EDITOR.getCode();");
+        if (codeContent.includes("Test passed!") && codeContent.length === 111) hasLoadedContent = true;
```
If these are the condition we are checking for the test to pass, then can we move those to the "assert" section to AAA (Arrange-Act-Assert)?

```js
const codeContent = await page.evaluate("window.EDITOR.getCode();");
await page.close();

expect(codeContent).toHaveLength(111);
expect(codeContent).toContain("Test passed!");
```

microbit-rosslowe replied:
LGTM.

microbit-rosslowe replied:
Changed in https://github.com/microbit-rosslowe/PythonEditor/pull/9/commits/3dcdaf967f1b0b1ec45374b6e936c721b0e0781d

microbit-carlos commented on tests/spec/puppeteer-spec.js:
```diff
@@ -0,0 +1,98 @@
+        let fileInput = await page.$('[name="load-form-file-upload"]');
+        await fileInput.uploadFile('./src/makecode.hex');
+        await page.click('[value="Load"]');
+        await page.waitFor(2000);
```
Is there any way to avoid this wait loop?

microbit-rosslowe replied:
It's not immediately obvious to me how we could, this also ensures the dialog opens within a given time limit.

microbit-carlos commented on tests/spec/puppeteer-spec.js:
```diff
@@ -0,0 +1,98 @@
+        await page.goto("http://localhost:5000/editor.html");
+
+        let hasLoadedV101Hex = false;
+        page.on('dialog', async dialog => {
```
Is this necessary if what we are ultimately checking is the content loaded into the code editor?
Same applied to the v0.9 test.

microbit-rosslowe replied:
Changed in https://github.com/microbit-rosslowe/PythonEditor/pull/9/commits/3dcdaf967f1b0b1ec45374b6e936c721b0e0781d

microbit-carlos commented on tests/spec/puppeteer-spec.js:
```diff
@@ -0,0 +1,98 @@
+        await fileInput.uploadFile('./src/1.0.1.hex');
+        await page.click('[value="Load"]');
+        const codeContent = await page.evaluate("window.EDITOR.getCode();");
+        if (codeContent.includes("PASS1") && codeContent.length === 32) hasLoadedV101Hex = true;
```
As with the other test we can move these into the "assert" section directly.
Same applied to the v0.9 test.

microbit-rosslowe replied:
Changed in https://github.com/microbit-rosslowe/PythonEditor/pull/9/commits/3dcdaf967f1b0b1ec45374b6e936c721b0e0781d

microbit-carlos commented on tests/spec/puppeteer-spec.js:
```diff
@@ -0,0 +1,98 @@
+describe("An editor for MicroPython running at localhost.", function() {
+
+    beforeAll(async() => {
+        global.browser = await global.puppeteer.launch({headless: true, args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage']});
```
Could you add a small comment about why these flags are used?

microbit-rosslowe replied:
Added in https://github.com/microbit-rosslowe/PythonEditor/pull/9/commits/3dcdaf967f1b0b1ec45374b6e936c721b0e0781d

microbit-rosslowe committed:
Updates from feedback

- rearrange AAA
- remove dialogs from old hex tests
- clarify flags
- reduce dialog wait

microbit-carlos commented:
Great work Ross, this is good to go! Do you want to submit a PR upstream and get it running on azure pipelines?

microbit-rosslowe committed:
Use jQuery to get script name

microbit-rosslowe committed:
Get tests working in Azure

microbit-rosslowe committed:
Merge branch 'master' of https://github.com/bbcmicrobit/PythonEditor into puppeteer-testing

microbit-rosslowe committed:
Use 1.1.4 only in testing

microbit-rosslowe committed:
Split wait into ms

microbit-rosslowe committed:
Consistent quotes

microbit-carlos committed:
Move hex files to subdirectory, and minor tweaks.
